### ISSUE TITLE
Question: why doesn't ruff complain about CapsWords (PascalCase)?

### ISSUE THREAD

**Issue by tribunsky-kir** [astral-sh/ruff#15069]
[PEP 8](https://peps.python.org/pep-0008/) describes that:

> Class names should normally use the CapWords convention.

and that for [Method Names and Instance Variables](https://peps.python.org/pep-0008/#method-names-and-instance-variables)

> Use the function naming rules: lowercase with words separated by underscores as necessary to improve readability.

---

Short `example.py`:

```python
class MyClass:
    WhatIf: str
    WhySo: str
```

```bash
ruff check example.py --select ALL --ignore D
All checks passed!
```

```
ruff --version
ruff 0.8.3
```

**Comment by Avasam** [comment]
Read-as-written, I believe attributes defined at the class-level are not instance variables, but rather class variables (unless you're using a dataclass)

```py
class MyClass:
    class_var = "bar"
    def __init__(self):
        self.instance_var = "foo"
```

That being said, I have no idea if PEP-8 intended to cover class variables here, or type-annotations-only declaractions.

**Comment by tmke8** [comment]
There is [N815](https://docs.astral.sh/ruff/rules/mixed-case-variable-in-class-scope/) which says:

Bad:

```python
class MyClass:
    myVariable = "hello"
    another_variable = "world"
```

Good:

```python
class MyClass:
    my_variable = "hello"
    another_variable = "world"
```

At first glance, it's surprising that it doesn't exclude PascalCase, but maybe there is a reason?

**Comment by tribunsky-kir** [comment]
> Read-as-written, I believe attributes defined at the class-level are not instance variables, but rather class variables (unless you're using a dataclass)
> 
> class MyClass:
>     class_var = "bar"
>     def __init__(self):
>         self.instance_var = "foo"
> 
> That being said, I have no idea if PEP-8 intended to cover class variables here, or type-annotations-only declaractions.

Technically, I can't disagree, that it is class attribute (and I am not sure about explicit PEP-8 opinion on that). 

But [N815 ](https://docs.astral.sh/ruff/rules/mixed-case-variable-in-class-scope/) would complain on `mixedCase` in the absolutely same case, as mentioned above. Such behavior looks inconsistent and surprising to me.

For popular case like `pydantic` class attributes are more like _instance variables_.

**Comment by MichaReiser** [comment]
These are just my own notes from looking into this issue:

The difference here doesn't come from whether the variable is attributed. It correctly flags `myVariable` but it doesn't flag `MyVariable`. See [playground](https://play.ruff.rs/0b9709d7-90e8-491a-87c2-cac1e144e779)

This sounds very familiar to https://github.com/astral-sh/ruff/issues/14905

**Comment by MichaReiser** [comment]
I did search the pep8-naming repository for an explanation but couldn't find any. 

My best guess is that CapNames are allowed because flagging them would be in conflict with the PEP8 recommendation for [type variables](https://peps.python.org/pep-0008/#type-variable-names).

**Comment by Daverball** [comment]
@MichaReiser Another code pattern that could explain this are nested classes, where the nested class is swapped out using an assignment to a different class by a subclass. I've seen this pattern used in a few places over the years.

E.g. `webob.request.Request.ResponseClass` or `wtforms.Form.Config` come to mind, although in the latter case you're generally expected to always define a new nested class, rather than assign one. Although you can think of those as type variables too, since they're referring to a type.

I think we could come up with some heuristics for the given example with `AnnAssign` though. If the annotation doesn't refer to a ` type`-like, then using PascalCase is wrong. If it's a regular `Assignment`, then we don't know, without performing type inference.

**Comment by tmke8** [comment]
I also sometimes use this pattern:

```python
from enum import Enum, auto
from typing import TypeAlias

class LoggerOption(Enum):
    INFO = auto()
    WARN = auto()

class Logger:
    Option: TypeAlias = LoggerOption

    def __init__(self, option: Option):  # using the alias
        self.option = option
```

In another file:
```python
from logger import Logger  # only Logger needs to be imported

logger = Logger(Logger.Option.INFO)
```
